[DO-NOT-MERGE] MSYS2 v3.2.0 preview 1

dscho opened the pull request:
Since there is a test release out of Cygwin v3.2.0, I figured I could rebase the `msys2-3_1_7_release` patches in preparation for an `msys2-3_2_0-release`.

Here is the range-diff relative to `msys2-3_1_7_release`:

```
 1:  73cffe479e !  1:  5e285afcad Add MSYS2 triplet
    @@ config.guess: EOF
     +	echo x86_64-unknown-msys
     +	exit ;;
          prep*:SunOS:5.*:*)
    - 	echo powerpcle-unknown-solaris2"`echo "$UNAME_RELEASE"|sed -e 's/[^.]*//'`"
    + 	echo powerpcle-unknown-solaris2"$(echo "$UNAME_RELEASE"|sed -e 's/[^.]*//')"
      	exit ;;
     
      ## config.rpath ##
    @@ newlib/libm/machine/i386/configure: else
          lt_cv_dlopen="dlopen"
          lt_cv_dlopen_libs=
          ;;
    -
    - ## winsup/config.guess ##
    -@@ winsup/config.guess: EOF
    -     amd64:CYGWIN*:*:* | x86_64:CYGWIN*:*:*)
    - 	echo x86_64-unknown-cygwin
    - 	exit ;;
    -+    amd64:MSYS*:*:* | x86_64:MSYS*:*:*)
    -+	echo x86_64-unknown-msys
    -+	exit ;;
    -     p*:CYGWIN*:*)
    - 	echo powerpcle-unknown-cygwin
    - 	exit ;;
    -
    - ## winsup/configure ##
    -@@ winsup/configure: CXX
    - CXXFLAGS
    - CCC
    - CPP'
    --ac_subdirs_all='cygwin cygserver doc
    -+ac_subdirs_all='cygwin cygserver
    - utils'
    - 
    - # Initialize some variables set by options.
    -@@ winsup/configure: export CXX
    - 
    - 
    - 
    --subdirs="$subdirs cygwin cygserver doc"
    -+subdirs="$subdirs cygwin cygserver"
    - 
    - if test "x$with_cross_bootstrap" != "xyes"; then
    -     subdirs="$subdirs utils"
    -
    - ## winsup/configure.ac ##
    -@@ winsup/configure.ac: AC_LANG(C++)
    - 
    - AC_CYGWIN_INCLUDES
    - 
    --AC_CONFIG_SUBDIRS(cygwin cygserver doc)
    -+AC_CONFIG_SUBDIRS(cygwin cygserver)
    - if test "x$with_cross_bootstrap" != "xyes"; then
    -     AC_CONFIG_SUBDIRS([utils])
    - fi
    -
    - ## winsup/cygserver/configure ##
    -@@ winsup/cygserver/configure: export CXX
    - 
    - 
    - case "$with_cross_host" in
    --  ""|*cygwin*)
    -+  ""|*cygwin*|*msys*)
    -     all_host="all_host"
    -     install_host="install_host"
    -     ;;
    -
    - ## winsup/cygserver/configure.ac ##
    -@@ winsup/cygserver/configure.ac: AC_LANG(C++)
    - AC_CYGWIN_INCLUDES
    - 
    - case "$with_cross_host" in
    --  ""|*cygwin*)
    -+  ""|*cygwin*|*msys*)
    -     all_host="all_host"
    -     install_host="install_host"
    -     ;;
    -
    - ## winsup/cygwin/configure ##
    -@@ winsup/cygwin/configure: export CXX
    - 
    - 
    - case "$with_cross_host" in
    --  ""|*cygwin*)
    -+  ""|*cygwin*|*msys*)
    -     all_host="all_host"
    -     install_host="install_host"
    -     ;;
    -
    - ## winsup/cygwin/configure.ac ##
    -@@ winsup/cygwin/configure.ac: AC_LANG(C++)
    - AC_CYGWIN_INCLUDES
    - 
    - case "$with_cross_host" in
    --  ""|*cygwin*)
    -+  ""|*cygwin*|*msys*)
    -     all_host="all_host"
    -     install_host="install_host"
    -     ;;
36:  da111b9773 !  2:  780dd31a9c Fix msys library name in import libraries
    @@ Metadata
      ## Commit message ##
         Fix msys library name in import libraries
     
    -    Cygwin's speclib doesn't handle dashes or dots. So we specifically fix up all the import libraries that link against `msys_2_0.dll` to correctly link against `msys-2.0.dll`.
    +    Cygwin's speclib doesn't handle dashes or dots. However, we are about to
    +    rename the output file name from `cygwin1.dll` to `msys-2.0.dll`.
    +
    +    Let's preemptively fix up all the import libraries that would link
    +    against `msys_2_0.dll` to correctly link against `msys-2.0.dll` instead.
     
      ## winsup/cygwin/speclib ##
     @@ winsup/cygwin/speclib: while (<$nm_fd>) {
 2:  929a7c5fcb !  3:  9b4339bfa8 Rename dll from cygwin to msys
    @@ winsup/Makefile.in: endif
     +	  ${INSTALL} $$i $(DESTDIR)$(prefix)/share/doc/Msys ; \
      	done
      
    - install: Makefile $(INSTALL_LICENSE) $(INSTALL_SUBDIRS)
    + install: Makefile install-license $(INSTALL_SUBDIRS)
     
      ## winsup/cygserver/Makefile.in ##
    -@@ winsup/cygserver/Makefile.in: LIBOBJS:=${patsubst %.o,lib%.o,$(OBJS)}
    - 
    - CYGWIN_OBJS:=$(cygwin_build)/version.o
    +@@ winsup/cygserver/Makefile.in: OBJS:=	cygserver.o client.o process.o msg.o sem.o shm.o threaded_queue.o \
    + 	sysv_msg.o sysv_sem.o sysv_shm.o setpwd.o pwdgrp.o
    + LIBOBJS:=${patsubst %.o,lib%.o,$(OBJS)}
      
     -CYGWIN_LIB:=$(cygwin_build)/libcygwin.a
     +CYGWIN_LIB:=$(cygwin_build)/libmsys-2.0.a
    @@ winsup/cygwin/Makefile.in: RUNTEST = `if [ -f $${srcdir}/../dejagnu/runtest ] ;
     -# built cygwin.dll and attempt to use it instead of the old one).
     +# built msys.dll and attempt to use it instead of the old one).
      
    - DLL_NAME:=@DLL_NAME@
    --TEST_DLL_NAME:=${patsubst %1.dll,%0.dll,@DLL_NAME@}
    +-DLL_NAME:=cygwin1.dll
    +-TEST_DLL_NAME:=${patsubst %1.dll,%0.dll,$(DLL_NAME)}
     -TEST_LIB_NAME:=libcygwin0.a
    --STATIC_LIB_NAME:=libcygwin_s.a
    ++DLL_NAME:=msys-2.0.dll
     +TEST_DLL_NAME:=${patsubst %-2.0.dll,%0.dll,@DLL_NAME@}
     +TEST_LIB_NAME:=libmsys0.a
    -+STATIC_LIB_NAME:=libmsys2_s.a
      DIN_FILE=@DIN_FILE@ common.din
     -DEF_FILE:=cygwin.def
     +DEF_FILE:=msys.def
    @@ winsup/cygwin/Makefile.in: maintainer-clean: clean
     +# Rule to build msys-2.0.dll
      $(TEST_DLL_NAME): $(LDSCRIPT) dllfixdbg $(DLL_OFILES) $(LIBSERVER) $(LIBC) $(LIBM) $(API_VER) Makefile $(VERSION_OFILES)
      	$(CXX) $(CXXFLAGS) \
    - 	-mno-use-libstdc-wrappers -L${WINDOWS_LIBDIR} \
    - 	-Wl,--gc-sections $(nostdlib) -Wl,-T$(firstword $^) -static \
    + 	-mno-use-libstdc-wrappers \
    + 	-Wl,--gc-sections -nostdlib -Wl,-T$(firstword $^) -static \
     -	-Wl,--heap=0 -Wl,--out-implib,cygdll.a -shared -o $@ \
     +	-Wl,--heap=0 -Wl,--out-implib,msysdll.a -shared -o $@ \
      	-e $(DLL_ENTRY) $(DEF_FILE) $(DLL_OFILES) $(VERSION_OFILES) \
    @@ winsup/cygwin/Makefile.in: maintainer-clean: clean
     -	${srcdir}/mkimport ${toolopts} ${NEW_FUNCTIONS} $@ cygdll.a $(wordlist 2,99,$^)
     +	${srcdir}/mkimport ${toolopts} ${NEW_FUNCTIONS} $@ msysdll.a $(wordlist 2,99,$^)
      
    - ${STATIC_LIB_NAME}: mkstatic ${TEST_DLL_NAME}
    --	perl -d $< -x ${EXCLUDE_STATIC_OFILES} --library=${LIBC} --library=${LIBM} --ar=${AR} $@ cygwin.map
    -+	perl -d $< -x ${EXCLUDE_STATIC_OFILES} --library=${LIBC} --library=${LIBM} --ar=${AR} $@ msys.map
    - 
      # Rule to make stub library used by testsuite
      # dependency set to $(LIB_NAME) to accommodate make -j2.
      $(TEST_LIB_NAME): $(LIB_NAME)
    @@ winsup/cygwin/common.din: cygwin_attach_handle_to_fd SIGFE
      cygwin_logon_user SIGFE
      cygwin_posix_path_list_p NOSIGFE
     
    - ## winsup/cygwin/configure ##
    -@@ winsup/cygwin/configure: fi
    - 
    - case "$target_cpu" in
    -    i?86)
    --		DLL_NAME="cygwin1.dll"
    -+		DLL_NAME="msys-2.0.dll"
    - 		DLL_ENTRY="_dll_entry@12"
    - 		DEF_DLL_ENTRY="dll_entry@12"
    - 		DIN_FILE="i686.din"
    - 		TLSOFFSETS_H="tlsoffsets.h"
    - 		;;
    -    x86_64)
    --		DLL_NAME="cygwin1.dll"
    -+		DLL_NAME="msys-2.0.dll"
    - 		DLL_ENTRY="dll_entry"
    - 		DEF_DLL_ENTRY="dll_entry"
    - 		DIN_FILE="x86_64.din"
    -
    - ## winsup/cygwin/configure.ac ##
    -@@ winsup/cygwin/configure.ac: dnl fi
    - 
    - case "$target_cpu" in
    -    i?86)
    --		DLL_NAME="cygwin1.dll"
    -+		DLL_NAME="msys-2.0.dll"
    - 		DLL_ENTRY="_dll_entry@12"
    - 		DEF_DLL_ENTRY="dll_entry@12"
    - 		DIN_FILE="i686.din"
    - 		TLSOFFSETS_H="tlsoffsets.h"
    - 		;;
    -    x86_64)
    --		DLL_NAME="cygwin1.dll"
    -+		DLL_NAME="msys-2.0.dll"
    - 		DLL_ENTRY="dll_entry"
    - 		DEF_DLL_ENTRY="dll_entry"
    - 		DIN_FILE="x86_64.din"
    -
      ## winsup/cygwin/crt0.c ##
     @@ winsup/cygwin/crt0.c: details. */
      
    @@ winsup/cygwin/dtable.cc: handle_to_fn (HANDLE h, char *posix_fn)
      	  if (w32len
     
      ## winsup/cygwin/exceptions.cc ##
    -@@ winsup/cygwin/exceptions.cc: try_to_debug (bool waitloop)
    +@@ winsup/cygwin/exceptions.cc: try_to_debug ()
        PWCHAR rawenv = GetEnvironmentStringsW () ;
        for (PWCHAR p = rawenv; *p != L'\0'; p = wcschr (p, L'\0') + 1)
          {
    @@ winsup/cygwin/x86_64.din
      EXPORTS
      #Exported variables
     
    - ## winsup/lsaauth/cyglsa.c ##
    -@@ winsup/lsaauth/cyglsa.c: LsaApLogonUserEx (PLSA_CLIENT_REQUEST request, SECURITY_LOGON_TYPE logon_type,
    - 	  return stat;
    - 	}
    - 
    --      memcpy (ts.SourceName, "Cygwin.1", 8);
    -+      memcpy (ts.SourceName, "MSYS.2", 6);
    -       ts.SourceIdentifier.HighPart = 0;
    -       ts.SourceIdentifier.LowPart = 0x0104;
    -       RtlInitEmptyUnicodeString (&flatnm, flatname,
    -
      ## winsup/testsuite/Makefile.in ##
     @@ winsup/testsuite/Makefile.in: ifdef VERBOSE
          RUNTESTFLAGS = -v
    @@ winsup/testsuite/config/default.exp
      proc winsup_version {} {
          global env
          global rootme
    --    clone_output "\n[exec grep ^%%% $rootme/../cygwin/cygwin0.dll]\n"
    +-    clone_output "\n[exec grep -a ^%%% $rootme/../cygwin/cygwin0.dll]\n"
     -    if { [info exists env(CYGWIN)] } {
     -        clone_output "CYGWIN=$env(CYGWIN)\n"
    -+    clone_output "\n[exec grep ^%%% $rootme/../cygwin/msys0.dll]\n"
    ++    clone_output "\n[exec grep -a ^%%% $rootme/../cygwin/msys0.dll]\n"
     +    if { [info exists env(MSYS)] } {
     +        clone_output "MSYS=$env(MSYS)\n"
          } else {
    @@ winsup/testsuite/config/default.exp
     
      ## winsup/testsuite/cygrun.c ##
     @@ winsup/testsuite/cygrun.c: main (int argc, char **argv)
    -       exit (0);
    -     }
    +   if (argc >= 3)
    +     SetEnvironmentVariable ("TDIRECTORY", argv[2]);
      
     -  SetEnvironmentVariable ("CYGWIN_TESTING", "1");
     -  if ((p = getenv ("CYGWIN")) == NULL || (strstr (p, "ntsec") == NULL))
    @@ winsup/testsuite/winsup.api/winsup.exp: foreach src [lsort [glob -nocomplain $sr
          } else {
     -	ws_spawn "$CC -nodefaultlibs -mwin32 $CFLAGS $src $add_includes $add_libs $runtime_root/binmode.o -lgcc $runtime_root/libcygwin0.a -lkernel32 -luser32 -o $base.exe"
     +	ws_spawn "$CC -nodefaultlibs -mwin32 $CFLAGS $src $add_includes $add_libs $runtime_root/binmode.o -lgcc $runtime_root/libmsys0.a -lkernel32 -luser32 -o $base.exe"
    - 	if { $rv != "" } {
    - 	    verbose -log "$rv"
    + 	if { $rv } {
      	    fail "$testcase (compile)"
    + 	} else {
     
      ## winsup/utils/Makefile.in ##
     @@ winsup/utils/Makefile.in: EXEEXT_FOR_BUILD:=@EXEEXT_FOR_BUILD@
      
      LDLIBS := -lnetapi32 -ladvapi32 -lkernel32 -luser32
    - CYGWIN_LDFLAGS := -static -Wl,--enable-auto-import -L${WINDOWS_LIBDIR} $(LDLIBS)
    + CYGWIN_LDFLAGS := -static -Wl,--enable-auto-import $(LDLIBS)
     -DEP_LDLIBS := $(cygwin_build)/libcygwin.a
     +DEP_LDLIBS := $(cygwin_build)/libmsys-2.0.a
      
    @@ winsup/utils/Makefile.in: cygcheck.exe: MINGW_LDFLAGS += ${ZLIB} -lwininet -lpsa
      strace.exe: MINGW_LDFLAGS += -lntdll
      
      ldd.exe:CYGWIN_LDFLAGS += -lpsapi -lntdll
    -@@ winsup/utils/Makefile.in: install: all
    - 	  $(INSTALL_PROGRAM) $$i $(DESTDIR)$(bindir)/$$n; \
    - 	done
    - 
    --$(cygwin_build)/libcygwin.a: $(cygwin_build)/Makefile
    -+$(cygwin_build)/libmsys-2.0.a: $(cygwin_build)/Makefile
    - 	@$(MAKE) -C $(@D) $(@F)
    - 
    - .PHONY: warn_dumper
     
      ## winsup/utils/cygcheck.cc ##
     @@ winsup/utils/cygcheck.cc: static const char *known_env_vars[] = {
 3:  0e31ff61a7 !  4:  b3e74fb72f Add functionality for converting UNIX paths in arguments and environment variables to Windows form for native Win32 applications.
    @@ Commit message
     
      ## winsup/cygwin/Makefile.in ##
     @@ winsup/cygwin/Makefile.in: DLL_OFILES:= \
    + 	mmap.o \
      	mmap_alloc.o \
      	msg.o \
    - 	msgcat.o \
     +	msys2_path_conv.o \
      	mount.o \
      	net.o \
    @@ winsup/cygwin/spawn.cc: child_info_spawn::worker (const char *prog_arg, const ch
      	      moreinfo->argc = newargv.argc;
      	      moreinfo->argv = newargv;
      	    }
    --	  if ((wincmdln || !real_path.iscygexec ())
    --	       && !cmd.fromargv (newargv, real_path.get_win32 (),
    --				 real_path.iscygexec ()))
     +	  else
    - 	    {
    --	      res = -1;
    --	      __leave;
    ++	    {
     +	      for (int i = 0; i < newargv.argc; i++)
     +	        {
     +	          //convert argv to win32
    @@ winsup/cygwin/spawn.cc: child_info_spawn::worker (const char *prog_arg, const ch
     +	          newargv.replace (i, tmpbuf);
     +	          free (tmpbuf);
     +	        }
    -+	      if ((wincmdln || !real_path.iscygexec ())
    -+	            && !cmd.fromargv (newargv, real_path.get_win32 (),
    -+			        real_path.iscygexec ()))
    -+	        {
    -+	          res = -1;
    -+	          __leave;
    -+	        }
    - 	    }
    - 
    - 
    ++	    }
    + 	  if ((wincmdln || !real_path.iscygexec ())
    +-	       && !cmd.fromargv (newargv, real_path.get_win32 (),
    +-				 real_path.iscygexec ()))
    ++	        && !cmd.fromargv (newargv, real_path.get_win32 (),
    ++		    real_path.iscygexec ()))
    + 	    {
    + 	      res = -1;
    + 	      __leave;
     @@ winsup/cygwin/spawn.cc: child_info_spawn::worker (const char *prog_arg, const char *const *argv,
            moreinfo->envp = build_env (envp, envblock, moreinfo->envc,
      				  real_path.iscygexec (),
 4:  8c4a21abdd =  5:  c2a9783f04 Add functionality for changing OS name via MSYSTEM environment variables.
 5:  d340694fa1 =  6:  6e09e92a17 - Move root to /usr. - Change sorting mount points. - By default mount without ACLs. - Can read /etc/fstab with short mount point format.
 6:  d771eaf777 <  -:  ---------- Do not create cygwin symlinks. Instead use deep copy of files/folders.
30:  48205ebc88 !  7:  d98873da68 squash! Do not create cygwin symlinks. Instead use deep copy of files/folders.
    @@
      ## Metadata ##
    -Author: Johannes Schindelin <johannes.schindelin@gmx.de>
    +Author: Алексей Павлов <alexey.pawlow@gmail.com>
     
      ## Commit message ##
    -    squash! Do not create cygwin symlinks. Instead use deep copy of files/folders.
    -
    -    The original patch to change the symbolic link behavior to create deep
    -    copies instead did so in a bit of a hacky way: it used the
    -    `WSYM_sysfile` code path, not introducing a separate mode. This makes it
    -    impossible for users to ask for Cygwin's behavior (and it also poses a
    -    maintenance burden: for example, we had to specifically disable Cygwin's
    -    code to prefer to create WSL-style symlinks whenever possible instead of
    -    using the original Cygwin hack to use special-formed system files).
    -
    -    Proposed new commit message when rebasing the MSYS2 patches on top of
    -    the next Cygwin runtime version:
    -
         Instead of creating Cygwin symlinks, use deep copy by default
     
         The new `winsymlinks` mode `deepcopy` (which is made the default) lets
    @@ winsup/cygwin/globals.cc: enum winsym_t
     @@ winsup/cygwin/globals.cc: bool ignore_case_with_glob;
      bool pipe_byte;
      bool reset_com;
    - bool wincmdln = true;
    + bool wincmdln;
     -winsym_t allow_winsymlinks = WSYM_sysfile;
     +winsym_t allow_winsymlinks = WSYM_deepcopy;
    - bool disable_pcon = true;
    + bool disable_pcon;
      
      bool NO_COPY in_forkee;
     
      ## winsup/cygwin/path.cc ##
    +@@ winsup/cygwin/path.cc: conv_path_list (const char *src, char *dst, size_t size,
    + 
    + /********************** Symbolic Link Support **************************/
    + 
    ++/*
    ++  Create a deep copy of src as dst, while avoiding descending in origpath.
    ++*/
    ++static int
    ++recursiveCopy (char * src, char * dst, const char * origpath)
    ++{
    ++  WIN32_FIND_DATA dHfile;
    ++  HANDLE dH;
    ++  BOOL findfiles;
    ++  int srcpos = strlen (src);
    ++  int dstpos = strlen (dst);
    ++  int res = -1;
    ++
    ++  debug_printf("recursiveCopy (%s, %s)", src, dst);
    ++
    ++  /* Create the destination directory */
    ++  if (!CreateDirectoryEx (src, dst, NULL))
    ++    {
    ++      debug_printf("CreateDirectoryEx(%s, %s, 0) failed", src, dst);
    ++      __seterrno ();
    ++      goto done;
    ++    }
    ++  /* Descend into the source directory */
    ++  if (srcpos + 2 >= MAX_PATH || dstpos + 1 >= MAX_PATH)
    ++    {
    ++      set_errno (ENAMETOOLONG);
    ++      goto done;
    ++    }
    ++  strcat (src, "\\*");
    ++  strcat (dst, "\\");
    ++  dH = FindFirstFile (src, &dHfile);
    ++  debug_printf("dHfile(1): %s", dHfile.cFileName);
    ++  findfiles = FindNextFile (dH, &dHfile);
    ++  debug_printf("dHfile(2): %s", dHfile.cFileName);
    ++  findfiles = FindNextFile (dH, &dHfile);
    ++  while (findfiles)
    ++    {
    ++      /* Append the directory item filename to both source and destination */
    ++      int filelen = strlen (dHfile.cFileName);
    ++      debug_printf("dHfile(3): %s", dHfile.cFileName);
    ++      if (srcpos + 1 + filelen >= MAX_PATH ||
    ++          dstpos + 1 + filelen >= MAX_PATH)
    ++        {
    ++          set_errno (ENAMETOOLONG);
    ++          goto done;
    ++        }
    ++      strcpy (&src[srcpos+1], dHfile.cFileName);
    ++      strcpy (&dst[dstpos+1], dHfile.cFileName);
    ++      debug_printf("%s -> %s", src, dst);
    ++      if (dHfile.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
    ++        {
    ++          /* Recurse into the child directory */
    ++          debug_printf("%s <-> %s", src, origpath);
    ++          if (strcmp (src, origpath)) // avoids endless recursion
    ++            if (recursiveCopy (src, dst, origpath))
    ++              goto done;
    ++        }
    ++      else
    ++        {
    ++          /* Just copy the file */
    ++          if (!CopyFile (src, dst, FALSE))
    ++            {
    ++              __seterrno ();
    ++              goto done;
    ++            }
    ++        }
    ++      findfiles = FindNextFile (dH, &dHfile);
    ++    }
    ++  if (GetLastError() != ERROR_NO_MORE_FILES)
    ++    {
    ++      __seterrno ();
    ++      goto done;
    ++    }
    ++  res = 0;
    ++
    ++done:
    ++
    ++  return res;
    ++}
    ++
    + /* Create a symlink from FROMPATH to TOPATH. */
    + 
    + extern "C" int
     @@ winsup/cygwin/path.cc: symlink_worker (const char *oldpath, path_conv &win32_newpath, bool isdevice)
      	}
            else
      	{
    --      path_conv src_path;
    --      src_path.check (oldpath, PC_SYM_NOFOLLOW, stat_suffixes);
    --      if (src_path.error)
    --        {
    --           set_errno (src_path.error);
    --           __leave;
    --        }
    --      if (!src_path.isdevice () && !src_path.is_fs_special ())
    --        {
    --           /* MSYS copy file instead make symlink */
    --
    --           char * real_oldpath;
    --           if (isabspath (oldpath))
    --             strcpy (real_oldpath = tp.c_get (), oldpath);
    --           else
    --              /* Find the real source path, relative
    --                 to the directory of the destination */
    --             {
    --                /* Determine the character position of the last path component */
    --                const char *newpath = win32_newpath.get_posix();
    --                int pos = strlen (newpath);
    --                while (--pos >= 0)
    --                  if (isdirsep (newpath[pos]))
    --                    break;
    --                /* Append the source path to the directory
    --                   component of the destination */
    --                if (pos+1+strlen(oldpath) >= MAX_PATH)
    --                  {
    --                     set_errno(ENAMETOOLONG);
    --                     __leave;
    --                  }
    --                strcpy (real_oldpath = tp.c_get (), newpath);
    --                strcpy (&real_oldpath[pos+1], oldpath);
    --             }
    --
    --           /* As a MSYS limitation, the source path must exist. */
    --		   path_conv win32_oldpath;
    --           win32_oldpath.check (real_oldpath, PC_SYM_NOFOLLOW, stat_suffixes);
    --           if (!win32_oldpath.exists ())
    --             {
    --                set_errno (ENOENT);
    --                __leave;
    --             }
    --
    --           char *w_newpath;
    --           char *w_oldpath;
    --           stpcpy (w_newpath = tp.c_get (), win32_newpath.get_win32());
    --           stpcpy (w_oldpath = tp.c_get (), win32_oldpath.get_win32());
    --           if (win32_oldpath.isdir())
    --             {
    --                char *origpath;
    --                strcpy (origpath = tp.c_get (), w_oldpath);
    --                res = recursiveCopy (w_oldpath, w_newpath, origpath);
    --             }
    --           else
    --             {
    --                if (!CopyFile (w_oldpath, w_newpath, FALSE))
    --                  {
    --                     __seterrno ();
    --                  }
    --                else
    --                  {
    --                     res = 0;
    --                  }
    --             }
    --           __leave;
    --        }
     +          if (wsym_type == WSYM_deepcopy)
     +	    {
     +	      path_conv src_path;
    @@ winsup/cygwin/path.cc: symlink_worker (const char *oldpath, path_conv &win32_new
     +		  __leave;
     +		}
     +	    }
    - 
    ++
      	  /* Default technique creating a symlink. */
      	  buf = tp.t_get ();
    + 	  cp = stpcpy (buf, SYMLINK_COOKIE);
 7:  6642d47553 =  8:  96fffd259c Automatically rewrite TERM=msys to TERM=cygwin With MSys1, it was necessary to set the TERM variable to "msys". To allow for a smooth transition from MSys1 to MSys2, let's simply handle TERM=msys as if the user had not specified TERM at all and wanted us to use our preferred TERM value.
 8:  42ee23210f =  9:  9c689b0d20 Do not convert environment for strace
 9:  87b48dc564 = 10:  2e2cf84044 Special case for converting root directory to have training slash
10:  8e4513fdb9 = 11:  445af9b2e3 dcrt0.cc: Untangle allow_glob from winshell
11:  9c98e9a830 = 12:  ad340ce966 dcrt0.cc (globify): Don't quote literal strings differently when dos_spec
12:  8e4f827dfa = 13:  daf4806c24 strace.cc: Don't set MSYS=noglob
13:  40d67da3ee = 14:  e8277973fc Add debugging for build_argv
14:  73a99bcd0d = 15:  7f7b58922a Add debugging for strace make_command_line
15:  c13ec33841 = 16:  e517273f7f environ.cc: New facility/environment variable MSYS2_ENV_CONV_EXCL
16:  adf8d56d40 = 17:  487c75df67 Fix native symbolic link spawn passing wrong arg0
17:  384223db8e = 18:  17caa0e0fd QueryUnbiasedInterruptTime must be load from kernel32.dll
18:  9312cda13a <  -:  ---------- fixup! Do not create cygwin symlinks. Instead use deep copy of files/folders.
19:  1f883dc43e = 19:  81c9b40918 strace --quiet: be *really* quiet
20:  69451787eb ! 20:  3a685c09fd Default to `disable_pcon`
    @@ winsup/cygwin/globals.cc
     @@ winsup/cygwin/globals.cc: bool pipe_byte;
      bool reset_com;
      bool wincmdln;
    - winsym_t allow_winsymlinks = WSYM_sysfile;
    + winsym_t allow_winsymlinks = WSYM_deepcopy;
     -bool disable_pcon;
     +bool disable_pcon = true;
      
21:  cf71f996e8 = 21:  bad433d179 Introduce the `enable_pcon` value for `MSYS`
22:  44bb133448 <  -:  ---------- Don't create WSL symlinks for MSYS2
23:  1bf51203be = 22:  1ebe04381e popen: call /usr/bin/sh instead of /bin/sh
24:  676936ee97 = 23:  b68e22d705 CI: add a GHA for doing a basic build test
35:  fa13d627f5 = 24:  41612d5fe6 CI: Upload the build artifacts
25:  2cdc58ccc9 = 25:  36e1bcda19 Set up a GitHub Action to keep in sync with Cygwin
26:  9557a16ab4 = 26:  8eee301eaf recursiveCopy: Always close the used handle with FindClose()
27:  7d384ab131 <  -:  ---------- fixup! Add functionality for converting UNIX paths in arguments and environment variables to Windows form for native Win32 applications.
28:  2b2cff60c6 ! 27:  c547c66007 Expose full command-lines to other Win32 processes by default
    @@ winsup/cygwin/globals.cc: bool allow_glob = true;
      bool reset_com;
     -bool wincmdln;
     +bool wincmdln = true;
    - winsym_t allow_winsymlinks = WSYM_sysfile;
    + winsym_t allow_winsymlinks = WSYM_deepcopy;
      bool disable_pcon = true;
      
     
29:  cfe1db3994 <  -:  ---------- Cygwin: pty: Prevent garbled output when pseudo console is disabled.
31:  f31004f660 <  -:  ---------- fixup! Don't create WSL symlinks for MSYS2
32:  65e26f2d5d <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=parentheses
33:  42babc8c84 <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=narrowing
34:  e68ccdf6e5 <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=narrowing
```

Explanations:

>  1:  73cffe479e !  1:  5e285afcad Add MSYS2 triplet

This huge interdiff is caused by dc93f7ef2d93dee6e0a44fb0caf37387f38e8611

>36:  da111b9773 !  2:  780dd31a9c Fix msys library name in import libraries

I moved this patch before the big DLL rename patch, and reworded the commit message to reflect that this is a preemptive patch now.

>  2:  929a7c5fcb !  3:  9b4339bfa8 Rename dll from cygwin to msys

This is also fallout from dc93f7ef2d93dee6e0a44fb0caf37387f38e8611, but also from 22d79c79b5ebe398f16ec33f9687c6dfeb511b4a

>  3:  0e31ff61a7 !  4:  b3e74fb72f Add functionality for converting UNIX paths in arguments and environment variables to Windows form for native Win32 applications.

This is 7d384ab131  squashed in.

>  6:  d771eaf777 <  -:  ---------- Do not create cygwin symlinks. Instead use deep copy of files/folders.
> 30:  48205ebc88 !  7:  d98873da68 squash! Do not create cygwin symlinks. Instead use deep copy of files/folders.

This is essentially d771eaf777 rewritten to _not_ override `WSYM_sysfile`, but to implement `WSYM_deepcopy` as a separate alternative.

> 22:  44bb133448 <  -:  ---------- Don't create WSL symlinks for MSYS2
> 31:  f31004f660 <  -:  ---------- fixup! Don't create WSL symlinks for MSYS2

This drops the change that disabled WSL symlinks: this is now guarded by the `WSYM_deepcopy` mode, and in `WSYM_sysfile` mode WSL symlinks will be heeded.

> 29:  cfe1db3994 <  -:  ---------- Cygwin: pty: Prevent garbled output when pseudo console is disabled.

The ConPTY support was completely overhauled in v3.2.0 (🤞 that it helps more than it breaks 😄 )

> 32:  65e26f2d5d <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=parentheses
> 33:  42babc8c84 <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=narrowing
> 34:  e68ccdf6e5 <  -:  ---------- Cygwin: avoid GCC 10 error with -Werror=narrowing

These were backports and have been integrated into v3.2.0.

mingwandroid committed:
Add debugging for strace make_command_line

mingwandroid committed:
Add debugging for build_argv

mingwandroid committed:
dcrt0.cc: Untangle allow_glob from winshell

Otherwise if globbing is allowed and we get called from a
Windows program, build_argv thinks we've been called from
a Cygwin program.

mingwandroid committed:
strace.cc: Don't set MSYS=noglob

Commit message for this code was:

* strace.cc (create_child): Set CYGWIN=noglob when starting new process so that

  Cygwin will leave already-parsed the command line alonw."

I can see no reason for it and it badly breaks the ability to use
strace.exe to investigate calling a Cygwin program from a Windows
program, for example:
strace mingw32-make.exe
.. where mingw32-make.exe finds sh.exe and uses it as the shell.
The reason it badly breaks this use-case is because dcrt0.cc depends
on globbing to happen to parse commandlines from Windows programs;
irrespective of whether they contain any glob patterns or not.

See quoted () comment:
"This must have been run from a Windows shell, so preserve
 quotes for globify to play with later."

mingwandroid committed:
dcrt0.cc (globify): Don't quote literal strings differently when dos_spec

Reverts 25ba8f306f3099caf8397859019e936b90510e8d. I can't figure out what
the intention was. I'm sure I'll find out soon enough when everything breaks.

This change means that input of:
  '"C:/test.exe SOME_VAR=\"literal quotes\""'

becomes:
  'C:/test.exe SOME_VAR="literal quotes"'

instead of:
  'C:/test.exe SOME_VAR=\literal quotes\'

.. which is at least consistent with the result for:
  '"no_drive_or_colon SOME_VAR=\"literal quotes\""'

The old result of course resulted in the quoted string being split into
two arguments at the space which is clearly not intended.

I *guess* backslashes in dos paths may have been the issue here?
If so I don't care since we should not use them, ever, esp. not at
the expense of sensible forward-slash-containing input.

mingwandroid committed:
environ.cc: New facility/environment variable MSYS2_ENV_CONV_EXCL

Works very much like MSYS2_ARG_CONV_EXCL. In fact it uses the same
function, arg_heuristic_with_exclusions (). Also refactors parsing
the env. variables to use new function, string_split_delimited ().

The env. that is searched through is the merged (POSIX + Windows)
one. It remains to be seen if this should be made an option or not.

This feature was prompted because the R language (Windows exe) calls
bash to run configure.win, which then calls back into R to read its
config variables (LOCAL_SOFT) and when this happens, msys2-runtime
converts R_ARCH from "/x64" to an absolute Windows path and appends
it to another absolute path, R_HOME, forming an invalid path.

dscho committed:
strace --quiet: be *really* quiet

The biggest problem with strace spitting out `create_child: ...` despite
being asked to be real quiet is that its output can very well interfere
with scripts' operations.

For example, when running any of Git for Windows' shell scripts with
`GIT_STRACE_COMMANDS=/path/to/logfile` (which is sadly an often needed
debugging technique while trying to address the many MSYS2 issues Git for
Windows faces), any time the output of any command is redirected into a
variable, it will include that `create_child: ...` line, wreaking havoc
with Git's expectations.

So let's just really be quiet when we're asked to be quiet.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

SquallATF committed:
Fix native symbolic link spawn passing wrong arg0

Alexpux committed:
Add MSYS2 triplet

Alexpux committed:
Rename dll from cygwin to msys

Alexpux committed:
Add functionality for converting UNIX paths in arguments and environment variables to Windows form for native Win32 applications.

Alexpux committed:
Add functionality for changing OS name via MSYSTEM environment variables.

Alexpux committed:
- Move root to /usr. - Change sorting mount points. - By default mount without ACLs. - Can read /etc/fstab with short mount point format.

Alexpux committed:
Instead of creating Cygwin symlinks, use deep copy by default

The new `winsymlinks` mode `deepcopy` (which is made the default) lets
calls to `symlink()` create (deep) copies of the source file/directory.

This is necessary because unlike Cygwin, MSYS2 does not try to be its
own little ecosystem that lives its life separate from regular Win32
programs: the latter have _no idea_ about Cygwin-emulated symbolic links
(i.e. system files whose contents start with `!<symlink>\xff\xfe` and
the remainder consists of the NUL-terminated, UTF-16LE-encoded symlink
target).

To support Cygwin-style symlinks, the new mode `sysfile` is introduced.

Co-authored-by: Johannes Schindelin <johannes.schindelin@gmx.de>

Alexpux committed:
Automatically rewrite TERM=msys to TERM=cygwin With MSys1, it was necessary to set the TERM variable to "msys". To allow for a smooth transition from MSys1 to MSys2, let's simply handle TERM=msys as if the user had not specified TERM at all and wanted us to use our preferred TERM value.

Alexpux committed:
Do not convert environment for strace

Strace is a Windows program so MSYS2 will convert all arguments and environment vars and that makes debugging msys2 software with strace very tricky.

Alexpux committed:
Special case for converting root directory to have training slash

Alexpux committed:
QueryUnbiasedInterruptTime must be load from kernel32.dll

dscho committed:
Set up a GitHub Action to keep in sync with Cygwin

This will help us by automating an otherwise tedious task.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Default to `disable_pcon`

The pseudo console support is just a bit too buggy still:

- When typing anything in an interactive Bash session before the prompt
  is shown, frequently the keystrokes are then replayed _twice_ when the
  prompt is active.

  Even worse: it seems that under certain circumstances (e.g. when
  spawning `less.exe` from a MINGW process), keystrokes are recorded
  while a process is active that wants to consume them but those
  recorded keystrokes are then replayed once the process finished (e.g.
  the `q` keystroke to exit `less.exe` will then be misinterpreted for
  interactive input in the Bash session).

- When `vim` is called from a MINGW process, it seems that the terminal
  loses the `onlcr` property after the `vim` process finished, i.e.
  subsequently printed lines do not start at the beginning of the line
  anymore, but precisely where the previous line ended.

- In `vim`, when selecting text visually (via the `v` keystroke), it
  seems that the selection is frequently reset while navigating with the
  arrow keys.

There are probably quite a few more rough edges in the pseudo console
feature, unfortunately.

In light of these issues, it makes most sense to disable the pseudo
console support and make it opt-in rather than opt-out.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Introduce the `enable_pcon` value for `MSYS`

It is simply the negation of `disable_pcon`, i.e. `MSYS=enable_pcon` is
equivalent to `MSYS=nodisable_pcon` (the former is slightly more
intuitive than the latter) and likewise `MSYS=noenable_pcon` is
equivalent to `MSYS=disable_pcon` (here, the latter is definitely more
intuitive than the former).

This is needed because we just demoted the pseudo console feature to be
opt-in instead of opt-out, and it would be awkward to recommend to users
to use "nodisable_pcon"... "nodisable" is not even a verb.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

lazka committed:
popen: call /usr/bin/sh instead of /bin/sh

We mount /usr/bin to /bin, but in a chroot this is broken and we
have no /bin, so try to use the real path.

chroot is used by pacman to run install scripts when called with --root
and this broke programs in install scripts calling popen()
(install-info from texinfo for example)

There are more paths hardcoded to /bin in cygwin which might also be broken
in this scenario, so this maybe should be extended to all of them.

lazka committed:
recursiveCopy: Always close the used handle with FindClose()

See https://github.com/msys2/MSYS2-packages/issues/2060

lazka committed:
CI: add a GHA for doing a basic build test

dscho committed:
Expose full command-lines to other Win32 processes by default

In the Cygwin project, it was decided that the command-line of Cygwin
processes, as shown in the output of `wmic process list`, would suffer
from being truncated to 32k (and is transmitted to the child process via
a different mechanism, anyway), and therefore only the absolute path of
the executable is shown by default.

Users who would like to see the full command-line (even if it is
truncated) are expected to set `CYGWIN=wincmdln` (or, in MSYS2's case,
`MSYS=wincmdln`).

Seeing as MSYS2 tries to integrate much better with the surrounding
Win32 ecosystem than Cygwin, it makes sense to turn this on by default.

Users who wish to suppress it can still set `MSYS=nowincmdln`.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

ktbarrett committed:
Fix msys library name in import libraries

Cygwin's speclib doesn't handle dashes or dots. However, we are about to
rename the output file name from `cygwin1.dll` to `msys-2.0.dll`.

Let's preemptively fix up all the import libraries that would link
against `msys_2_0.dll` to correctly link against `msys-2.0.dll` instead.

lazka committed:
CI: Upload the build artifacts

dscho commented:
I fear we'll have to disable Cygwin's GitHub workflows... They seem not to work.

dscho commented:
> 
> 
> I fear we'll have to disable Cygwin's GitHub workflows... They seem not to work.

Huh. https://github.com/rpmsphere/noarch/tree/master/r only contains https://github.com/rpmsphere/noarch/blob/master/r/rpmsphere-release-32-1.noarch.rpm, not what Cygwin's workflow expects (https://github.com/rpmsphere/noarch/raw/master/r/rpmsphere-release-33-1.noarch.rpm)...

dscho commented:
Sadly, I made a mistake in resolving the merge conflicts, so I needed https://github.com/msys2/msys2-runtime/pull/30/commits/9fb9c2212319da858c0b4ac0e2fa591352b1dc06

Also, it seems that our `mingw-w64-headers` version is older than Cygwin expects (and it recently learned to assume incorrectly what's available in v8.0.0 of that package), so we have to revert an upstream commit for now.

Finally, I managed to work around one problem in the `cygwin` workflow only to run into more of them.

lazka commented:
I've updated mingw-w64 here: https://github.com/msys2/MSYS2-packages/pull/2374 so will be available in the not-so-distant future.

lazka commented:
oh, or is that using msys2-w32api-headers..hmm

dscho commented:
> 
> 
> oh, or is that using msys2-w32api-headers..hmm

It is.

lazka commented:
ok, will do later.

lazka commented:
We should now have a recent mingw-w64 everywhere.

dscho commented:
I finally had time to figure out why the `cygwin` builds fail. Guess [what](https://github.com/msys2/msys2-runtime/pull/30/checks?check_run_id=2116832652#step:5:21):

>  mingw64-headers                noarch     7.0.0-2.fc33       fedora      4.8 M

They cannot even guarantee their minimum requirement in their own CI builds. Which is apparently not a problem for them? The builds indeed [all fail](https://github.com/cygwin/cygwin/actions)...

Any objections against simply deleting their dysfunctional GitHub workflow in the `msys2-*` branches?

lazka commented:
Fine with me.

What about changing https://github.com/msys2/msys2-runtime/blob/d10d0d9b03f858b10fa4b9a5dd0dd0c5364ca439/.github/workflows/cygwin.yml#L3 from "push" to "workflow_dispatch", so it's easier to rebase and it can still be triggered manually if needed.

dscho committed:
Disable the 'cygwin' GitHub workflow

It does not work at all. For example, `rpm -E %fedora` says that there
should be version 33 of rpmsphere at
https://github.com/rpmsphere/noarch/tree/master/r, but there is only
version 32.

Another thing that is broken: Cygwin now assumes that a recent
mingw-w64-headers version is available, but Fedora apparently only
offers v7.0.0, which is definitely too old to accommodate for the
expectation of https://github.com/cygwin/cygwin/commit/c1f7c4d1b6d7.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
> What about changing
> 
> https://github.com/msys2/msys2-runtime/blob/d10d0d9b03f858b10fa4b9a5dd0dd0c5364ca439/.github/workflows/cygwin.yml#L3

Done.

lazka commented:
Once thing I've noticed is that `iscygpty` like used here https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-python/1700-cygpty-isatty.patch seems to no longer work (python has to be started with "-i" again to get interactive mode)

lazka commented:
https://cygwin.com/pipermail/cygwin/2021-March/248091.html

dscho commented:
> 
> 
> I've asked about it upstream: https://cygwin.com/pipermail/cygwin/2021-March/248091.html

@lazka https://cygwin.com/pipermail/cygwin-patches/2021q1/011260.html is supposed to fix that, correct?

dscho commented:
> 
> 
> > I've asked about it upstream: https://cygwin.com/pipermail/cygwin/2021-March/248091.html
> 
> @lazka https://cygwin.com/pipermail/cygwin-patches/2021q1/011260.html is supposed to fix that, correct?

@lazka FYI I applied and pushed it. Once the build finishes, may I bother you to test this with `ptycheck`?

lazka commented:
I'll check tomorrow

lazka commented:
Looks good, and python is happy again:

```
Is stdin a tty? no
Is stdout a tty? no
Is stderr a tty? no

0 (Pipe): \msys-dd50a72ab4668b33-pty0-from-master-nat, is_cygpty: 1
1 (Pipe): \msys-dd50a72ab4668b33-pty0-to-master-nat, is_cygpty: 1
2 (Pipe): \msys-dd50a72ab4668b33-pty0-to-master-nat, is_cygpty: 1

Is running on mintty (or other term)? yes
```

dscho commented:
Yeah!

BTW to resolve merge conflicts, I rebased the patches, resolving the fixups (and dropping the no-longer necessary commits). The range-diff relative to the pre-rebase state:

```diff
34:  d56bfed2fe =  1:  ef05e8bdae Cygwin: pty: Rename input/output named pipes.
 -:  ---------- >  2:  2533912fc7 Cygwin: Allow executing Windows Store's "app execution aliases"
 -:  ---------- >  3:  0631c6644e Cygwin: Treat Windows Store's "app execution aliases" as symbolic links
 -:  ---------- >  4:  f578883232 Cygwin: add "app execution aliases" treatment to release notes
 1:  5e285afcad =  5:  2dc8548f51 Add MSYS2 triplet
 2:  780dd31a9c =  6:  5e23b7ebff Fix msys library name in import libraries
 3:  9b4339bfa8 !  7:  1902cb394e Rename dll from cygwin to msys
    @@ winsup/cygwin/Makefile.in: RUNTEST = `if [ -f $${srcdir}/../dejagnu/runtest ] ;
     -TEST_DLL_NAME:=${patsubst %1.dll,%0.dll,$(DLL_NAME)}
     -TEST_LIB_NAME:=libcygwin0.a
     +DLL_NAME:=msys-2.0.dll
    -+TEST_DLL_NAME:=${patsubst %-2.0.dll,%0.dll,@DLL_NAME@}
    ++TEST_DLL_NAME:=${patsubst %-2.0.dll,%0.dll,$(DLL_NAME)}
     +TEST_LIB_NAME:=libmsys0.a
      DIN_FILE=@DIN_FILE@ common.din
     -DEF_FILE:=cygwin.def
 4:  b3e74fb72f =  8:  85834517a2 Add functionality for converting UNIX paths in arguments and environment variables to Windows form for native Win32 applications.
 5:  c2a9783f04 =  9:  c6eca09866 Add functionality for changing OS name via MSYSTEM environment variables.
 6:  6e09e92a17 = 10:  926e501ea1 - Move root to /usr. - Change sorting mount points. - By default mount without ACLs. - Can read /etc/fstab with short mount point format.
 7:  d98873da68 = 11:  96c67015dd Instead of creating Cygwin symlinks, use deep copy by default
 8:  96fffd259c = 12:  610a2a4fcd Automatically rewrite TERM=msys to TERM=cygwin With MSys1, it was necessary to set the TERM variable to "msys". To allow for a smooth transition from MSys1 to MSys2, let's simply handle TERM=msys as if the user had not specified TERM at all and wanted us to use our preferred TERM value.
 9:  9c689b0d20 = 13:  9b66fa0548 Do not convert environment for strace
10:  2e2cf84044 = 14:  ad60f22743 Special case for converting root directory to have training slash
11:  445af9b2e3 = 15:  85d4de44b9 dcrt0.cc: Untangle allow_glob from winshell
12:  ad340ce966 = 16:  fd56b3d6f9 dcrt0.cc (globify): Don't quote literal strings differently when dos_spec
13:  daf4806c24 = 17:  95fda37063 strace.cc: Don't set MSYS=noglob
14:  e8277973fc = 18:  ec875f8531 Add debugging for build_argv
15:  7f7b58922a = 19:  f9eb3fb673 Add debugging for strace make_command_line
16:  e517273f7f = 20:  d9df8ec515 environ.cc: New facility/environment variable MSYS2_ENV_CONV_EXCL
17:  487c75df67 = 21:  b216b825c9 Fix native symbolic link spawn passing wrong arg0
18:  17caa0e0fd = 22:  c91decd4ba QueryUnbiasedInterruptTime must be load from kernel32.dll
19:  81c9b40918 = 23:  c00fa53325 strace --quiet: be *really* quiet
20:  3a685c09fd = 24:  e50db40edf Default to `disable_pcon`
21:  bad433d179 = 25:  d60979423b Introduce the `enable_pcon` value for `MSYS`
22:  1ebe04381e = 26:  dd3e1a9429 popen: call /usr/bin/sh instead of /bin/sh
23:  b68e22d705 = 27:  366703088d CI: add a GHA for doing a basic build test
24:  41612d5fe6 = 28:  4606a4f01a CI: Upload the build artifacts
25:  36e1bcda19 = 29:  2eda2605cf Set up a GitHub Action to keep in sync with Cygwin
26:  8eee301eaf = 30:  60d92c8d5f recursiveCopy: Always close the used handle with FindClose()
27:  c547c66007 = 31:  5f5bb4bf6a Expose full command-lines to other Win32 processes by default
28:  47708be62c <  -:  ---------- Revert "Cygwin: winlean.h: remove most of the extended memory API"
29:  9fb9c22123 <  -:  ---------- fixup! Rename dll from cygwin to msys
30:  81e5d3f314 = 32:  c31abe5e77 Disable the 'cygwin' GitHub workflow
31:  ac98b2f631 <  -:  ---------- fixup! Revert "Cygwin: winlean.h: remove most of the extended memory API"
32:  11e5616aa9 <  -:  ---------- Cygwin: pty: Rename input named pipes.
33:  c1ef8584e2 <  -:  ---------- fixup! Cygwin: pty: Rename input named pipes.
```

As you can see, the "Cygwin: pty: Rename input named pipes." patch made it upstream, the winlean.h revert is gone, and the fixup to "Rename dll from cygwin to msys" was squashed in.

lazka commented:
3.2.0 is out:

* https://cygwin.com/git/?p=newlib-cygwin.git;a=commit;h=f578883232d578bd83132f1f94a7f6826a74048e
* https://cygwin.com/pipermail/cygwin/2021-March/248137.html

dscho commented:
> 
> 
> 3.2.0 is out:
> 
>     * https://cygwin.com/git/?p=newlib-cygwin.git;a=commit;h=f578883232d578bd83132f1f94a7f6826a74048e
> 
>     * https://cygwin.com/pipermail/cygwin/2021-March/248137.html

Excellent. And even better: the tagged version is precisely the revision onto which I had rebased this PR.

Therefore, I pushed the branch directly to `msys2-3_2_0-release`.

dscho commented:
@lazka here is the MSYS2-packages Pull Request: https://github.com/msys2/MSYS2-packages/pull/2414

dscho commented:
And with this, I'll close the Pull Request (which was never meant to be merged into `cygwin/master` 😁)...
